model_api/python/model_api/models/visual_prompting.py

```diff
@@ -0,0 +1,675 @@
+            },
+        )
+
+        for prompt in processed_prompts:
+            label = prompt.pop("label")
+            prompt.update(**image_embeddings)
+
+            prediction = self.decoder_model.infer_sync(prompt)
+            prediction["scores"] = prediction["iou_predictions"]
+            prediction["labels"] = label
+            processed_prediction = self.decoder_model.postprocess(prediction, meta)
+            outputs.append(processed_prediction)
+
+        return VisualPromptingResult(
+            upscaled_masks=[item["upscaled_masks"] for item in outputs],
+            low_res_masks=[item["low_res_masks"] for item in outputs],
+            iou_predictions=[item["iou_predictions"] for item in outputs],
+            scores=[item["scores"] for item in outputs],
+            labels=[item["labels"] for item in outputs],
+            hard_predictions=[item["hard_prediction"] for item in outputs],
+            soft_predictions=[item["soft_prediction"] for item in outputs],
+        )
+
+    def __call__(
+        self,
+        image: np.ndarray,
+        boxes: Prompt | None = None,
+        points: Prompt | None = None,
+    ) -> VisualPromptingResult:
+        return self.infer(image, boxes, points)
+
+
+class SAMLearnableVisualPrompter:
+    def __init__(
+        self,
+        encoder_model: SAMImageEncoder,
+        decoder_model: SAMDecoder,
+        reference_features: VisualPromptingFeatures | None = None,
+    ):
+        self.encoder_model = encoder_model
+        self.decoder_model = decoder_model
+
+        if reference_features is not None:
+            self._reference_features = reference_features.feature_vectors
+            self._used_indices = reference_features.used_indices
+        else:
+            self._reference_features = None
+            self._used_indices = None
+
+        self._point_labels_box = np.array([[2, 3]], dtype=np.float32)
+        self._has_mask_inputs = [np.array([[0.0]]), np.array([[1.0]])]
+
+        self._is_cascade: bool = False
+        self._threshold: float = 0.0
+        self._num_bg_points: int = 1
+        self._default_threshold_target: float = 0.65
+        self._image_size: int = self.encoder_model.image_size
+        self._downsizing: int = 64
+        self._default_threshold_reference: float = 0.3
+
+    def has_reference_features(self) -> bool:
+        return self._reference_features is not None and self._used_indices is not None
+
+    @property
+    def reference_features(self) -> VisualPromptingFeatures:
+        if self.has_reference_features():
+            return VisualPromptingFeatures(np.copy(self._reference_features), np.copy(self._used_indices))
+
+        raise RuntimeError("Reference features are not generated")
+
+    def learn(
+        self,
+        image: np.ndarray,
+        boxes: Prompt | None = None,
+        points: Prompt | None = None,
+        reset_features: bool = False
+    ) -> tuple[VisualPromptingFeatures, np.ndarray]:
```

Comment:
How is learning on multiple images/labels handled?

Reply:
Zero-shot learning is done on each image individually. Learn() returns the reference features vectors for each label. Labels are contained in prompts. Later, the reference feature vectors can be used to obtain object masks on other images.

Currently, learning the same classes on multiple images is not considered, since first we need to prove that feature vector arithmetic somehow works in that scenario.
However, consequent learn() calls still make sense if we pass new prompts with unseen labels. In that case, a list of reference features would be extended with the ones corresponding to new labels.